Expanded texture APIs

penguinencounter opened the pull request:
4P5's feature request: [Thread](https://discord.com/channels/1129805506354085959/1274653529943838811)
DogeIsCut's feature request: [Thread](https://discord.com/channels/1129805506354085959/1329305436213481635)

## Feature List

- [x] `Texture:multiply(Texture, x, y, w, h)`, `:add`, `:subtract`, `:divide`
- [x] `Texture:resize(outputName, x, y`~~`, mode`~~`)`
- [x] `Texture:invert(x, y, w, h, invertAlpha?)`
- [x] `Texture:blit(options)`
- [x] ~~`Texture:setWrapMode(mode)`~~ `Texture:setOverflowMode(mode)`
    - [x] `error` (default)
    - [x] `ignore` (aka `discard`)
    - [x] `wrap`
    - [x] `mirror`
    - [x] `clamp`
- [x] `Texture:getOverflowMode()`
- [x] `Texture:line(x1, y1, x2, y2, color)`
- [x] `Texture:remapColors(from, to)`
- [x] `Texture:applyMatrix`: remove `clip`
- [x] add `getWidth(): int`, `getHeight(): int` to Lua allowlist

## Bug fixes too
- [x] Clamp input color vectors between 0-1 to prevent overflowing into other channels
- [x] `fill` diverges between 1.20 and 1.21.4 (probably 1.21.3)

## Finishing touches
- [x] Documentation
- [x] Test for edge cases and weird combinations of things (also how the 'restore' function handles these)

## Live [GitHub Actions](https://github.com/penguinencounter/Figura/actions/workflows/build.yml?query=branch%3Amoretextures) (1.20)
All branches (cherry-pick): [Multitarget build](https://github.com/penguinencounter/Figura/actions/runs/13362679839)
most branches which have diverged significantly from 1.20 have merge conflicts, sorry paladin

penguinencounter committed:
Expanded texture APIs (WIP)

penguinencounter committed:
Additional guards against overflowing colors

penguinencounter committed:
`:invert(x, y, w, h, alpha?)` implementation

penguinencounter committed:
bump

penguinencounter committed:
downgrade switch to support 1.16.5 cherrypick

penguinencounter committed:
partial flexmat and flexvec implementation just to see if this will crossbuild at all

penguinencounter committed:
Fixes and Java 8 compatibility

penguinencounter committed:
full matrix implementation

penguinencounter committed:
add CLAMP

penguinencounter committed:
All abstract methods covered

penguinencounter committed:
Conversion methods

penguinencounter committed:
internal generic image filtering

penguinencounter committed:
resize

penguinencounter committed:
Clean incompatible imports

penguinencounter committed:
Reduce scope

- remove applyFilter because of performance issues and general 'who would ever use this'
- remove FlexibleMatrix and FlexibleVector because they are now unused and maybe another PR should implement them instead

penguinencounter committed:
add getWidth and getHeight to lua api

why not?

penguinencounter committed:
blit()

penguinencounter committed:
Lines, normal (works great) and antialiased (good enough)

moves & renames BlendMode to be outside BlitOptions
contains cursed dispatcher. someone please redesign @LuaWhitelist to support overloads thanks

penguinencounter commented:
Features done, docs still WIP and real-use testing should probably be done

penguinencounter committed:
Remove unused

penguinencounter committed:
o7 argument line handler

penguinencounter committed:
quick consistency fix, more hell tomorrow

penguinencounter committed:
documented it!

penguinencounter committed:
fix 1.16.5 java8 compat

penguinencounter committed:
fix 1.16.5 java8 compat part 2

UnlikePaladin commented:
mfw when testing

UnlikePaladin commented:
do mark it as ready when it's ready cause like otherwise how will I know

penguinencounter commented:
look, for this one I did test all of the features, i don't have like a full avatar that uses a bunch of them at the same time in weird edge-case ways

UnlikePaladin commented:
oh okay, just mark it as ready for review when you think it's ready to be merged

penguinencounter commented:
i want to throw pallete swapping functionality on top of this because it's in the same general area before it gets merged if that's okay with you

(there's a separate feature request for this)

UnlikePaladin commented:
Okay, that's fine by me 👍🏻

penguinencounter committed:
Fix logical issues with backing up images

penguinencounter committed:
Fix issue with not mixing colors when the blend mode might require it

penguinencounter committed:
Fix me being dumb

penguinencounter committed:
Rename WriteOverflowStrategy to TextureOverflowStrategy

penguinencounter committed:
Remove unused code

penguinencounter committed:
Don't throw on `clip` and ignore instead (is this better?)

penguinencounter committed:
optimize: use Actual Position getset

(in applyFunc/applyMatrix where we already have the Actual Position from the Virtual Position)

penguinencounter committed:
feat: remapColors (nodoc, untested)

also i can't get my dev env to work right now hold on

penguinencounter commented:
![Screenshot 2025-03-17 at 22-50-45 Showcase](https://github.com/user-attachments/assets/ead0dfed-a35b-4475-946c-a7a64a85f2e6)

penguinencounter commented:
@UnlikePaladin i think this is good to go for review 😅

penguinencounter committed:
docs: remap_colors

UnlikePaladin commented:
wait this is finally ready? 💀 
![titanic](https://github.com/user-attachments/assets/c1247958-77e1-45c2-a8ba-01a1468badaa)

UnlikePaladin reviewed (approved):
lgtm

penguinencounter commented:
@UnlikePaladin
![6 months ago](https://github.com/user-attachments/assets/56e25e72-b33c-4b2d-8b3f-07986b183835)

UnlikePaladin commented:
tbh it got buried under other PRs
